### PR TITLE
SDSTOR-8752 mapping pba implementation

### PR DESCRIPTION
**Pull request by yamingk**
This PR contains the impmentation for below apis:
1. try_map_pba, 
2. update_map_pba, 
3. remove_map_pba
4. async_fetch_write_pbas

### CODE REVIEW

File: src/lib/state_machine/state_machine.cpp [hunk 1]
```diff
@@ -185,39 +194,123 @@ repl_req* ReplicaStateMachine::lsn_to_req(int64_t lsn) {
...
-    return std::make_pair(local_pbas_ptr->pbas, local_pbas_ptr->state);
+
+    return std::make_pair(local_pbas_ptr->m_pbas, local_pbas_ptr->m_state);
+}
+
+//
+// if return false /* no need to wait */, no cb will be triggered;
+// if return true /* need to wait */ , cb will be triggered after all local pbas completed writting;
+//
+bool ReplicaStateMachine::async_fetch_write_pbas(const std::vector< fully_qualified_pba >& fq_pba_list,
```
[the remark this answers is not in the document]
Comment (yamingk) [reply src/lib/state_machine/state_machine.cpp:218]: it depends on the common use case from the caller, it was called by ReplicaLogStore::end_of_append_batch, which I think most of cases will be more than 4 elements (which is not suited for small_vector).

File: src/lib/state_machine/state_machine.cpp [hunk 2]
```diff
@@ -185,39 +194,123 @@ repl_req* ReplicaStateMachine::lsn_to_req(int64_t lsn) {
 }
 
 std::pair< pba_list_t, pba_state_t > ReplicaStateMachine::try_map_pba(const fully_qualified_pba& fq_pba) {
-#if 0
-    const auto it = m_pba_map.find(fq_pba);
+    const auto it = m_pba_map.find(fq_pba.to_key_string());
     local_pba_info_ptr local_pbas_ptr{nullptr};
     if (it != m_pba_map.end()) {
```
[the remark this answers is not in the document]
Comment (yamingk) [reply src/lib/state_machine/state_machine.cpp:199]: Somehow I found it is more readable this way.

File: src/lib/state_machine/state_machine.cpp [hunk 3]
```diff
@@ -185,39 +194,123 @@ repl_req* ReplicaStateMachine::lsn_to_req(int64_t lsn) {
...
         local_pbas_ptr = it->second;
     } else {
         const auto local_pbas = m_state_store->alloc_pbas(fq_pba.size);
-        assert(local_pbas);
+        assert(local_pbas.size());
 
-        local_pbas_ptr = std::make_shared< local_pba_info >(local_pbas, pba_state_t::allocated,
-                                                            local_pbas.size() /* ref_cnt */, nullptr /*waiter*/);
+        local_pbas_ptr = std::make_shared< local_pba_info >(local_pbas, pba_state_t::allocated, nullptr /*waiter*/);
```
[the remark this answers is not in the document]
Comment (yamingk) [reply src/lib/state_machine/state_machine.cpp:205]: I see your point, somehow this way I can use the same variable to return the value which looks cleaner.

File: src/include/home_replication/repl_decls.h [hunk 4]
```diff
@@ -21,6 +23,7 @@ struct fully_qualified_pba {
     uint32_t server_id;
     pba_t pba;
     uint32_t size; // corresponding size of this pba;
+    std::string to_key_string() const { return fmt::format("{}_{}", std::to_string(server_id), std::to_string(pba)); }
```
Comment (raakella1) [inline src/include/home_replication/repl_decls.h:26]: We can have a static variable here and return that.
```
static const std::string key_string {fmt::format("{}_{}", std::to_string(server_id), std::to_string(pba))};
return key_string;
```

**Review by raakella1 — APPROVED** [review]
LG! minor nit

File: src/include/home_replication/repl_decls.h (hunk 4 above)
Comment (yamingk) [reply src/include/home_replication/repl_decls.h:26]: let me review this comment and resolve it in next PR

**Pull request merged**